src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java

```diff
@@ -2488,15 +2561,130 @@ interface DateTimePrinterParser {
+            return true;
+        }
+    }
+
+    static final class LocalTimePrinterParser extends CompositePrinterParser {
+        private final boolean nanoOptional;
+        private final int nanoMinWidth;
+        private final int nanoMaxWidth;
+        private final int fractionalDigits;
+
+        LocalTimePrinterParser(boolean nanoOptional, int nanoMinWidth, int nanoMaxWidth) {
+            super(buildPrinterParsers(nanoOptional, nanoMinWidth, nanoMaxWidth), false);
+            this.nanoOptional = nanoOptional;
+            this.nanoMinWidth = nanoMinWidth;
+            this.nanoMaxWidth = nanoMaxWidth;
+            if (nanoMaxWidth == 0 && nanoMaxWidth == 9) {
+                fractionalDigits = -2;
+            } else {
+                fractionalDigits = nanoMaxWidth;
+            }
+        }
+
+        private static List<DateTimePrinterParser> buildPrinterParsers(
+                boolean optional,
+                int nanoMinWidth,
+                int nanoMaxWidth
+        ) {
+            List<DateTimePrinterParser> printerParsers = new ArrayList<>();
+            printerParsers.add(new NumberPrinterParser(ChronoField.HOUR_OF_DAY, 2, 2, SignStyle.NOT_NEGATIVE));
+            printerParsers.add(new CharLiteralPrinterParser(':'));
+            printerParsers.add(new NumberPrinterParser(ChronoField.MINUTE_OF_HOUR, 2, 2, SignStyle.NOT_NEGATIVE));
+            printerParsers.add(new CharLiteralPrinterParser(':'));
+            printerParsers.add(new NumberPrinterParser(ChronoField.SECOND_OF_MINUTE, 2, 2, SignStyle.NOT_NEGATIVE));
+
+            if (nanoMinWidth >= 0) {
+                DateTimePrinterParser nanosPrinterParser
+                        = new NanosPrinterParser(nanoMinWidth, nanoMaxWidth, optional, 0);
+                if (optional) {
+                    nanosPrinterParser
+                            = new CompositePrinterParser(new DateTimePrinterParser[]{nanosPrinterParser}, true);
+                }
+                printerParsers.add(nanosPrinterParser);
+            }
+
+            return printerParsers;
+        }
+
+        @Override
+        public boolean format(DateTimePrintContext context, StringBuilder buf) {
+            TemporalAccessor temporal = context.getTemporal();
+
+            LocalTime time;
+            if (temporal instanceof LocalTime) {
+                time = (LocalTime) temporal;
+            } else if (temporal instanceof LocalDateTime) {
+                time = ((LocalDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof ZonedDateTime) {
+                time = ((ZonedDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetDateTime) {
+                time = ((OffsetDateTime) temporal).toLocalTime();
+            } else if (temporal instanceof OffsetTime) {
+                time = ((OffsetTime) temporal).toLocalTime();
+            } else {
+                return super.format(context, buf);
+            }
```

Comment:
This can use `instanceof <pattern>`:
```suggestion
            if (temporal instanceof LocalTime lt) {
                time = lt;
            } else if (temporal instanceof LocalDateTime ldt) {
                time = ldt.toLocalTime();
            } else if (temporal instanceof ZonedDateTime zdt) {
                time = zdt.toLocalTime();
            } else if (temporal instanceof OffsetDateTime odt) {
                time = odt.toLocalTime();
            } else if (temporal instanceof OffsetTime ot) {
                time = ot.toLocalTime();
            } else {
                return super.format(context, buf);
            }
```

Or even a pattern switch:
```suggestion
            switch (temporal) {
                case LocalTime lt -> time = lt;
                case LocalDateTime ldt -> time = ldt.toLocalTime();
                case ZonedDateTime zdt -> time = zdt.toLocalTime();
                case OffsetDateTime odt -> time = odt.toLocalTime();
                case OffsetTime ot -> time = ot.toLocalTime();
                default -> {
                    return super.format(context, buf);
                }
            }
```

Reply:
It's a good idea to use the new switch syntax